# VA opdracht 1

Vul hier je naam in en de naam van je groepsgenoot: 

## Deadline

De deadline voor deze opdracht is woensdag vóór 10:00.

We maken eerst gebruik van matplotlib. Hier kan je snel makkelijke plots mee maken. Ook gaan we aan het eind  gebruik maken van Seaborn. Dit werkt ietsje anders, maar kan je wel nog mooiere plots mee maken.

## Deel 1 - Schaatsen

We gaan kijken naar de 500m meter tijden van de broers Ronald en Michel Mulder geschaatst in Thialf (bron:osta.nl). De dataset is te vinden op BrightSpace / Teams. Zorg dat deze dataset zich in dezelfde map als deze notebook zit. Ter voorbereiding op een visualisatie nemen we de volgende stappen.
* We laden de package `pandas` in en noemen het `pd`.
* We gebruiken de functie `head` om de eerste vijf regels uit de dataset te bekijken.
* We gebruiken de functie `info` om de structuur van de dataset te bekijken.
* We gebruiken de functie `describe` om de statische waarden te bekijken.

In [1]:
import sys
import pandas as pd
schaatstijden = pd.read_csv('Schaatstijden500m.csv')
schaatstijden['Datum'] = pd.to_datetime(schaatstijden['Datum'])

In [2]:
schaatstijden.head()

Tijd      Datum                  Ijsbaan  Rijder
0  89.80 2020-01-26  IJsstadion Thialf (NED)  Ronald
1  34.86 2020-01-25  IJsstadion Thialf (NED)  Ronald
2  34.92 2020-01-18  IJsstadion Thialf (NED)  Ronald
3  35.27 2019-12-27  IJsstadion Thialf (NED)  Michel
4  35.38 2019-12-27  IJsstadion Thialf (NED)  Ronald

In [3]:
schaatstijden.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   Tijd     76 non-null     float64       
 1   Datum    76 non-null     datetime64[ns]
 2   Ijsbaan  76 non-null     object        
 3   Rijder   76 non-null     object        
dtypes: datetime64[ns](1), float64(1), object(2)
memory usage: 2.5+ KB


In [4]:
schaatstijden.describe()

Tijd
count  76.000000
mean   35.815263
std     6.280937
min    34.490000
25%    34.917500
50%    35.055000
75%    35.272500
max    89.800000

## Onderdeel 1

Kijk eens goed naar de tabellen hierboven en geef antwoord op de volgende vragen.
* Hoeveel 500m tijden zitten er in de dataset?
* Wat is de snelste tijd?

#### Antwoord:
* ant 1 3 
* ant  34.49 

## Onderdeel 2

Maak een histogram van de kolom `Tijd`. Vul de code aan, voldoe aan de volgende extra eisen en pas verder geen dingen aan.
* Kies de juiste kolom.
* Maak de histogram geel.
* Gebruik een geschikte titel.
* Gebruik een geschikt x-as label.
* Gebruik een geschikt y-as label.

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
plt.hist()
# plt.title()
# plt.xlabel()
# plt.ylabel()
plt.show()

Wellicht heb je het al gezien. Een van de broers is gevallen tijdens een rit. Dit resulteerde in tijd een stuk langzamer dan normaal. Filter deze tijd uit de dataset en maak de histogram nogmaals.

In [ ]:
#Code om te filteren


In [ ]:
plt.hist()

## Onderdeel 3

Voor de volgende analyse gaan we een boxplot maken. (gebruik de dataset zonder val) We willen een uitspraak doen welke broer sneller is op de 500m in Thialf. Maak een boxplot van de tijden van Ronald en een boxplot van de tijden van Michel. Voldoe aan de volgende eisen.
* Voeg een lijn blauwe toe met het huidige baan record in thialf, 34,07
* Gebruik een geschikte titel.
* Gebruik een geschikt x-as label.
* Gebruik een geschikt y-as label.

In [ ]:
schaatstijdenBoxplotData = {'Ronald': schaatstijden[schaatstijden['Rijder'] == 'Ronald']['Tijd'],
                            'Michel': schaatstijden[schaatstijden['Rijder'] == 'Michel']['Tijd']}

In [ ]:
schaatstijdenBoxplotData['Ronald'].describe()

In [ ]:
schaatstijdenBoxplotData['Michel'].describe()

In [ ]:
#code voor boxplot

fig, ax = plt.subplots()

ax.boxplot([schaatstijdenBoxplotData[...], schaatstijdenBoxplotData[...]])

# Add x-axis tick labels:


# Add a y-axis label


plt.show()

## Onderdeel 4

Deze vraag gaat over visualisatie gemaakt bij onderdeel 3. Doe een uitspraak over de welke broer het snelst is. Gebruik een maat voor locatie (waar is het gemiddelde?) en voor spreiding (liggen de verschillende tijden ver uit elkaar of dicht bij elkaar?) in je antwoord.

#### Antwoord

## Onderdeel 5

De broers hebben tijdens een aantal wedstrijden tegen elkaar gereden (niet uitsluitend in dezelfde rit). Ook deze data staat op BrightSpace, we laden de data in en berekenen het verschil van het onderlinge resultaat.

In [ ]:

import pandas as pd
import matplotlib.pyplot as plt
schaatstijden = pd.read_csv('/Users/jermaineveneman/Desktop/Schaatstijden500m verschil.csv')

schaatstijden['Verschil'] = schaatstijden['Ronald'] - schaatstijden['Michel']

plt.hist(schaatstijden['Verschil'], bins=10, color='skyblue', edgecolor='black')
plt.title('Histogram van verschil tussen Ronald en Michel')
plt.xlabel('Verschil (seconden)')
plt.ylabel('Aantal wedstrijden')
plt.grid(axis='y', alpha=0.75)
plt.show()

ronald_mean = schaatstijden['Ronald'].mean()
michel_mean = schaatstijden['Michel'].mean()

ronald_std = schaatstijden['Ronald'].std()
michel_std = schaatstijden['Michel'].std()

print("\nGemiddelde tijd Ronald:", round(ronald_mean, 2))
print("Gemiddelde tijd Michel:", round(michel_mean, 2))
print("Spreiding Ronald (standaardafwijking):", round(ronald_std, 2))
print("Spreiding Michel (standaardafwijking):", round(michel_std, 2))

if ronald_mean < michel_mean:
    print("\nConclusie: Ronald is gemiddeld sneller.")
elif michel_mean < ronald_mean:
    print("\nConclusie: Michel is gemiddeld sneller.")
else:
    print("\nConclusie: Beide zijn gemiddeld even snel.")

if ronald_std < michel_std:
    print("Ronald is consistenter in zijn tijden.")
else:
    print("Michel is consistenter in zijn tijden.")

In [ ]:
schaatstijdenVerschil.head()

Maakt een histogram van het verschil tussen de broers.

In [ ]:
plt.hist(schaatstijdenVerschil['Verschil'])
plt.title('Histogram van verschil tussen Ronald en Michel')
plt.xlabel('Verschil (seconden)')
plt.ylabel('Aantal wedstrijden')
plt.show()

In [ ]:
plt.hist(schaatstijdenVerschil['Verschil'])
plt.title('Histogram van verschil tussen Ronald en Michel')
plt.xlabel('Verschil (seconden)')
plt.ylabel('Aantal wedstrijden')
plt.show()



## Onderdeel 6

Deze vraag gaat over visualisatie gemaakt bij onderdeel 5. Doe een uitspraak over de welke broer het snelst is. Gebruik een maat voor locatie en voor spreiding in je antwoord.

#### Antwoord

In [ ]:
Gemiddelde tijd:
Ronald: 34.93 seconden
Michel: 35.19 seconden
Standaardafwijking:
Ronald: 0.22 seconden
Michel: 0.23 seconden
Conclusie:
Ronald is gemiddeld sneller dan Michel.
Ronald is iets consistenter (lagere standaardafwijking).

## Onderdeel 7

Dit is een iets moeilijkere vraag, voor als de voorgaande opdrachten redelijk makkelijk waren. Deze meer geavanceerde visualisatie gaat over het verloop van de tijden over de jaren. Maak een dataframe van schaatstijdenVerschil met de gemiddelde 500m tijd per broer per jaar (mocht je over de skills beschikken, dan mag het ook per schaatsseizoen september t/m maart).

In [ ]:
#code om gemiddelde per jaar te berekenen

In [ ]:
gemPerJaarPerBroer.head()

Maak een lijndiagram met op de x-as het jaar en op de y-as de gemiddelde tijd per jaar. Geef een lijn per broer weer. Geef de visualisatie geschikte titels en een legenda.

# Deel 2 - Seaborn package

### Voorbeeld

We gaan voor de laatste twee onderdelen gebruik maken van een andere package: Seaborn. Hier kan je redelijk makkelijk mooie plots maken. Het werkt wel net wat anders qua input en output. In de volgende twee onderdelen lopen we hier met je doorheen, zodat je volgende week bij de case visualisaties kan maken met Seaborn.

Mocht je bij het importeren een foutmelding krijgen, dan heb je waarschijnlijk nog niet de package geinstalleerd. Dit kan je doen door het hekje voor de eerste stukjes code weg te halen.

In [ ]:
#!conda install --yes --prefix {sys.prefix} numpy
#!conda install seaborn
import seaborn as sns

We gaan nu een histogram maken met een simpel voorbeeld (waarbij we andere data inladen). Een histogram maak je als volgt. Zie ook de voorbeelden: https://seaborn.pydata.org/generated/seaborn.histplot.html

In [ ]:
penguins = sns.load_dataset("penguins")
sns.histplot(data=penguins, x="flipper_length_mm")


## Onderdeel 8
Zorg nu dat je een histogram maakt voor de schaatstijden! waarbij op de x-as de tijden staan.

In [ ]:
sns.histplot(data=..., x = ...)

Maak een histogram waarbij je voor de verschillende schaatsers (Ronald, Michel) een andere kleur geeft. 

In [ ]:
sns.histplot(data=..., x = ..., hue = ...)

## Onderdeel 9

Met de vraag uit **onderdeel 7** gaan we ook kijken wat we nog meer kunnen zeggen over de huidige data. 
* Hoe vaak van alle keren was Ronald/Michel sneller?
* Hoeveel procent van de keren was dit?
* Maak een top 20 van snelste tijden van de langzamere broer en maak een top 20 van langzaamste tijden van de snellere broer.
* Hoe vaak had de langzame broer kunnen winnen van de snelle broer als hij zijn snelste tijden had gereden en de snelle broer zijn langzaamste tijden? (gebruik de gesoorteerde tabellen van de tijden om dit te tellen)


#### Antwoord